src/peer_select.cc

```diff
@@ -623,11 +623,8 @@ PeerSelector::selectSomeNeighbor()
 
             if (ping.n_replies_expected > 0) {
                 entry->ping_status = PING_WAITING;
-                eventAdd("PeerSelector::HandlePingTimeout",
-                         HandlePingTimeout,
-                         this,
-                         0.001 * ping.timeout,
-                         0);
+                auto *event = eventAdd("PeerSelector::HandlePingTimeout", HandlePingTimeout, this, 0.001 * ping.timeout, 0);
```

Comment:
AFAICT, this official code is pretty bad:

* It uses raw "this" for a cbdata-protected child class. Sooner or later, code like this leads to "real" bugs similar to those fixed in ce9bb79, (38a9f55, 9e64d84 where Amos should have used the commit message from my [patch](https://bugs.squid-cache.org/attachment.cgi?id=2647)), and 4299f87.
* It uses a static wrapper for an (essentially) AsyncCall-driven callback.

Both problems may disappear if we adjust event API to support AsyncCalls (i.e. `SQUID-521`), but since we are modifying this code _now_, we should ask ourselves whether these problems should be addressed now as well, avoiding repeated modifications of the same code. I see two primary ways to address these problems:

1. Do some of the `SQUID-521` work: Use a true AsyncCall callback for events like this. We already have cbdataDialer(). We will probably need to add a NullaryCbdataDialer (while avoiding code duplication with the existing UnaryCbdataDialer).
1. Stop abusing eventAdd() for transaction-specific ping timeouts -- the existing event-handling code is not optimized to support thousands of timeouts that can be created when thousands of concurrent transactions schedule timeouts. Add a global object that will schedule at most _one_ ping timeout at a time, maintaining an internal list of transactions that need this ping timeout service. That new service will handle code context correctly, of course. This is what existing ThePrimeChanceGiver and TheSpareAllowanceGiver do, for example (see HappyOrderEnforcer::startedWaiting()).

If your #78 classification and my other change requests in this PR are correct, then this eventAdd() will become the only remaining transaction-specific eventAdd(). With that assumption in mind, the second option becomes even more attractive!